### PR TITLE
Implement the job to acquire zip boundaries container from ftp server…

### PR DESCRIPTION
**Pull request by FranckLecuyer**

### CODE REVIEW

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                } else {
+                    filesImportFailed.add(fileName);
+                }
+            } else {
+                filesAlreadyImported.add(fileName);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } catch (InterruptedException e) {
+            throw e;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java:67]: Not necessary ?

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+                    importBoundary(boundaryContent, allBoundaryIds, fileName, cgmesBoundaryServiceRequester, filesImported, filesAlreadyImported, filesImportFailed);
+                }
+
+                entry = zis.getNextEntry();
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } catch (InterruptedException e) {
+            throw e;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java:103]: Not necessary ?

File: src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        // Closing boundary
+        byteArrays.add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8));
+
+        // Serializing as byte array
+        return HttpRequest.BodyPublishers.ofByteArrays(byteArrays);
+    }
+
+    public List<String> getBoundariesIds() throws IOException, InterruptedException {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java:94]: IOException not necessary

File: src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                for (int i = 0; i < array.length(); i++) {
+                    result.add(array.getString(i));
+                }
+                return result;
+            }
+        } catch (IOException e) {
+            LOGGER.error("I/O Error while getting all boundary ids");
+        } catch (InterruptedException e) {
+            LOGGER.error("Interruption while getting all boundary ids");
+            Thread.currentThread().interrupt();
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java:116]: Why this interrupt ?

File: src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        HttpRequest request = HttpRequest.newBuilder()
+                .uri(URI.create(serviceUrl + API_VERSION + "/boundaries"))
+                .timeout(Duration.ofMinutes(2))
+                .header("Content-Type", "multipart/form-data;boundary=" + boundary)
+                .POST(ofMimeMultipartData(data, boundary))
+                .build();
+
+        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+        return response.statusCode() == 200;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java:56]: 20x ?

File: src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                }
+            }
+        }
+        return false;
+    }
+
+    private static boolean isValidFileVersion(String version) {
+        try {
+            int v = Integer.parseInt(version);
+            return version.length() == 3 && v > 0 && v < 1000;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java:55]: Perhaps constants ?
Log a message with the version when length != 3 ?

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
+ */
+public class BoundaryAcquisitionServer implements AutoCloseable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BoundaryAcquisitionServer.class);
+
+    private final StandardFileSystemManager fsManager = new StandardFileSystemManager();
+    private FileSystemOptions fsOptions = new FileSystemOptions();
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:36]: final ?

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    public BoundaryAcquisitionServer(String url, String userName, String password) throws FileSystemException {
+        serverUrl = url;
+
+        StaticUserAuthenticator auth = new StaticUserAuthenticator(null, userName, password);
+        DefaultFileSystemConfigBuilder.getInstance().setUserAuthenticator(fsOptions, auth);
+
+        SftpFileSystemConfigBuilder.getInstance().setUserDirIsRoot(fsOptions, true);
+        SftpFileSystemConfigBuilder.getInstance().setPreferredAuthentications(fsOptions, "publickey,password");
+        SftpFileSystemConfigBuilder.getInstance().setConnectTimeoutMillis(fsOptions, 30000);
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:48]: Add a constant ?

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    public Map<String, String> listFiles(String acquisitionDirPath) throws IOException {
+        FileObject serverRoot = fsManager.resolveFile(serverUrl, fsOptions);
+        FileObject acquisitionDirectory = serverRoot.resolveFile(acquisitionDirPath);
+        List<FileObject> childrenFiles = Arrays.stream(acquisitionDirectory.getChildren()).filter(f -> {
+            try {
+                // filter on zip files that matches pattern
+                return f.isFile() && CgmesBoundaryUtils.isValidBoundaryContainerFileName(f.getName().getBaseName());
+            } catch (FileSystemException e) {
+                return false;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:67]: Warn message ?

**Comment by SlimaneAmar** [comment]
Did you investigate the causes of security hotspots ?

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+                                                   List<String> allBoundaryIds,
+                                                   CgmesBoundaryServiceRequester cgmesBoundaryServiceRequester,
+                                                   List<String> filesImported,
+                                                   List<String> filesAlreadyImported,
+                                                   List<String> filesImportFailed) throws InterruptedException {
+        String fileName;
+        try (SecuredZipInputStream zis = new SecuredZipInputStream(new ByteArrayInputStream(acquiredFile.getData()), CgmesBoundaryUtils.MAX_ZIP_ENTRIES_COUNT, CgmesBoundaryUtils.MAX_ZIP_SIZE)) {
+            ZipEntry entry = zis.getNextEntry();
+            while (entry != null) {
+                if (new File(entry.getName()).getCanonicalPath().contains("..")) {
```
Comment (JacquesBorsenberger) [inline src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java:81]: A file or directory can be named with more than one '.'  'like..this...' for example isn't  startsWith better ?

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java (hunk 8 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:48]: or a parameters, some server a slow to respond?

File: src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    private CgmesBoundaryUtils() {
+    }
+
+    /* The boundary zip container file name should have the following structure:
+    <effectiveDateTime>__ENTSOE_BD_<fileVersion>.zip
+    where :
+    <effectiveDateTime>: UTC datetime (YYYYMMDDTHHmmZ)
+    <fileVersion>: three characters long positive integer number between 000 and 999
+     */
+    private static final String DOT_REGEX = "\\.";
```
Comment (JacquesBorsenberger) [inline src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java:32]: I prerfer [.] for the dot regex, as it doesn't risks multiple evaluations

File: src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    private static final String UNDERSCORE_REGEX = "\\_";
```
Comment (JacquesBorsenberger) [inline src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java:33]: _ is not a special regEx character it shouldn't be escaped.

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java (hunk 1 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java]: Removed

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java (hunk 2 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java]: Removed

File: src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java:57]: Same test is done in cgmes-assembling-job and case-import-job
And 200 is the standard HTTP response code when everything is OK

File: src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/CgmesBoundaryServiceRequester.java:117]: To fix sonarcloud report :  "InterruptedException" should not be ignored

File: src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java (hunk 6 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java:55]: Done

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java (hunk 7 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:36]: Done

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java (hunk 8 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:48]: Constant added

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java (hunk 9 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionServer.java:67]: Done

**Comment by FranckLecuyer** [comment]
> Did you investigate the causes of security hotspots ?

These security hotspots are still returned by SonarCloud analysis, but there are safe, because we use a custom class SecuredZipInputStream to read the zip file

File: src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java (hunk 10 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/BoundaryAcquisitionJob.java]: Done

File: src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java (hunk 11 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java:32]: Done

File: src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java (hunk 12 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/boundary/importer/job/utils/CgmesBoundaryUtils.java:33]: Done

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+    <dependencies>
+        <!-- Compilation dependencies -->
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-vfs2</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-commons</artifactId>
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply pom.xml:157]: Removed

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply pom.xml:161]: Dependency removed

**Review by JacquesBorsenberger — APPROVED** [review]

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**